Running SPCON... 0x403E518F21F5Ce308085Dcf6637758C61f92446A 600 2022-05-23
Installing solc '0.4.21'...
Version '0.4.21' installed.
2022-05-23
{'limit': 60, 'network': 'ethereum', 'address': '0x403E518F21F5Ce308085Dcf6637758C61f92446A', 'date': '2022-05-23'}
0x403E518F21F5Ce308085Dcf6637758C61f92446A RC
./0x403E518F21F5Ce308085Dcf6637758C61f92446A
loaded abi.
3  functions ['withdraw', 'buyTokens', 'setTime']
48  users
Timecost for loading history: 0.01040196418762207
No.user: 48; No.func: 3
+----------------------------------------------------+
| Basic roles statistics (id, len(users), functions) |
+---------+--------+---------------------------------+
|  RoleId | Users  |            Functions            |
+---------+--------+---------------------------------+
|    0    |   47   |          ['buyTokens']          |
|    1    |   1    |     ['withdraw', 'setTime']     |
+---------+--------+---------------------------------+
Gen. 0 (0.00%): Max/Min/Avg Fitness(Raw)             [566.53(1000.00)/388.38(3.98)/472.11(472.11)]
Gen. 100 (100.00%): INFO:spcon.symExec:Totally 0 integrity policies
INFO:spcon.symExec:Totally 1 integrity policies
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 587, in symExec
    SymExecEVM().symCreateContractAccount(contract=contract, EthereumAddress=EthereumAddress)
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 473, in symCreateContractAccount
    contract_account = m.solidity_create_contract(source_code = address, \
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 643, in solidity_create_contract
    contract_account = self.create_contract(
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 730, in create_contract
    self._transaction("CREATE", owner, balance, address, data=init, gas=gas)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 991, in _transaction
    ) = self._migrate_tx_expressions(state, caller, address, value, data, gas, price)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 893, in _migrate_tx_expressions
    data = state.migrate_expression(data)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/state.py", line 458, in migrate_expression
    migrated_expression = self.constraints.migrate(expression, name_migration_map=migration_map)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/constraints.py", line 366, in migrate
    migrated_expression = replace(expression, object_migration_map)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 1043, in replace
    visitor.visit(expression, use_fixed_point=True)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 90, in visit
    value = self._method(node, *operands)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 63, in _method
    return self._rebuild(expression, args)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 116, in _rebuild
    aux = copy.copy(expression)
  File "/usr/lib/python3.8/copy.py", line 102, in copy
    return _reconstruct(x, None, *rv)
  File "/usr/lib/python3.8/copy.py", line 272, in _reconstruct
    y.__setstate__(state)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1070, in __setstate__
    array = array.store(index, value)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 832, in store
    return ArrayStore(array=self, index=self.cast_index(index), value=self.cast_value(value))
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1037, in __init__
    super().__init__(array=array, operands=(index, value), **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1016, in __init__
    self._operands = (array, *operands)
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 69, in handler
    _raise_exception(timeout_exception, exception_message)
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 45, in _raise_exception
    raise exception()
timeout_decorator.timeout_decorator.TimeoutError: 'Timed Out'
CRITICAL:spcon.symExec:RC-0x403E518F21F5Ce308085Dcf6637758C61f92446A cannot be deployed using manticore
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 587, in symExec
    SymExecEVM().symCreateContractAccount(contract=contract, EthereumAddress=EthereumAddress)
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 473, in symCreateContractAccount
    contract_account = m.solidity_create_contract(source_code = address, \
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 643, in solidity_create_contract
    contract_account = self.create_contract(
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 730, in create_contract
    self._transaction("CREATE", owner, balance, address, data=init, gas=gas)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 991, in _transaction
    ) = self._migrate_tx_expressions(state, caller, address, value, data, gas, price)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 893, in _migrate_tx_expressions
    data = state.migrate_expression(data)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/state.py", line 458, in migrate_expression
    migrated_expression = self.constraints.migrate(expression, name_migration_map=migration_map)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/constraints.py", line 366, in migrate
    migrated_expression = replace(expression, object_migration_map)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 1043, in replace
    visitor.visit(expression, use_fixed_point=True)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 90, in visit
    value = self._method(node, *operands)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 63, in _method
    return self._rebuild(expression, args)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 116, in _rebuild
    aux = copy.copy(expression)
  File "/usr/lib/python3.8/copy.py", line 102, in copy
    return _reconstruct(x, None, *rv)
  File "/usr/lib/python3.8/copy.py", line 272, in _reconstruct
    y.__setstate__(state)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1070, in __setstate__
    array = array.store(index, value)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 832, in store
    return ArrayStore(array=self, index=self.cast_index(index), value=self.cast_value(value))
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1037, in __init__
    super().__init__(array=array, operands=(index, value), **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1016, in __init__
    self._operands = (array, *operands)
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 69, in handler
    _raise_exception(timeout_exception, exception_message)
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 45, in _raise_exception
    raise exception()
timeout_decorator.timeout_decorator.TimeoutError: 'Timed Out'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/__main__.py", line 36, in execute_command
    SymExecEVM().fuzzing(contract=contractName, EthereumAddress=args.eth_address, separation_policies = separation_policies, integrity_policies=integrity_policies,\
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 82, in new_function
    return function(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 653, in fuzzing
    attack_test_sequence = greedyTesting(data, S=Stack, unauthorizedusers = unauthorizedusers, desc=f"Testing separation priviledgefuncs...\nR: {authorizedrole[1]}\n")
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 611, in greedyTesting
    if not symExec(unauthorizedusers, S):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 591, in symExec
    exit(-1)
  File "/usr/lib/python3.8/_sitebuiltins.py", line 26, in __call__
    raise SystemExit(code)
SystemExit: -1
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 587, in symExec
    SymExecEVM().symCreateContractAccount(contract=contract, EthereumAddress=EthereumAddress)
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 473, in symCreateContractAccount
    contract_account = m.solidity_create_contract(source_code = address, \
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 643, in solidity_create_contract
    contract_account = self.create_contract(
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 730, in create_contract
    self._transaction("CREATE", owner, balance, address, data=init, gas=gas)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 991, in _transaction
    ) = self._migrate_tx_expressions(state, caller, address, value, data, gas, price)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/ethereum/manticore.py", line 893, in _migrate_tx_expressions
    data = state.migrate_expression(data)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/state.py", line 458, in migrate_expression
    migrated_expression = self.constraints.migrate(expression, name_migration_map=migration_map)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/constraints.py", line 366, in migrate
    migrated_expression = replace(expression, object_migration_map)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 1043, in replace
    visitor.visit(expression, use_fixed_point=True)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 90, in visit
    value = self._method(node, *operands)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 63, in _method
    return self._rebuild(expression, args)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/visitors.py", line 116, in _rebuild
    aux = copy.copy(expression)
  File "/usr/lib/python3.8/copy.py", line 102, in copy
    return _reconstruct(x, None, *rv)
  File "/usr/lib/python3.8/copy.py", line 272, in _reconstruct
    y.__setstate__(state)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1070, in __setstate__
    array = array.store(index, value)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 832, in store
    return ArrayStore(array=self, index=self.cast_index(index), value=self.cast_value(value))
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1037, in __init__
    super().__init__(array=array, operands=(index, value), **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/manticore-0.3.8.dev231202-py3.8.egg/manticore/core/smtlib/expression.py", line 1016, in __init__
    self._operands = (array, *operands)
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 69, in handler
    _raise_exception(timeout_exception, exception_message)
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 45, in _raise_exception
    raise exception()
timeout_decorator.timeout_decorator.TimeoutError: 'Timed Out'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/__main__.py", line 36, in execute_command
    SymExecEVM().fuzzing(contract=contractName, EthereumAddress=args.eth_address, separation_policies = separation_policies, integrity_policies=integrity_policies,\
  File "/usr/local/lib/python3.8/dist-packages/timeout_decorator-0.5.0-py3.8.egg/timeout_decorator/timeout_decorator.py", line 82, in new_function
    return function(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 653, in fuzzing
    attack_test_sequence = greedyTesting(data, S=Stack, unauthorizedusers = unauthorizedusers, desc=f"Testing separation priviledgefuncs...\nR: {authorizedrole[1]}\n")
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 611, in greedyTesting
    if not symExec(unauthorizedusers, S):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/symExec.py", line 591, in symExec
    exit(-1)
  File "/usr/lib/python3.8/_sitebuiltins.py", line 26, in __call__
    raise SystemExit(code)
SystemExit: -1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/__main__.py", line 67, in main
    execute_command(args)
  File "/usr/local/lib/python3.8/dist-packages/SpCon-0.1.0-py3.8.egg/spcon/__main__.py", line 41, in execute_command
    raise Exception("Unkown error")
Exception: Unkown error
Max/Min/Avg Fitness(Raw)             [1001.63(1000.00)/0.00(3.98)/711.16(711.16)]
Total time elapsed: 4.881 seconds.
best role number: 2
Role#0:{'buyTokens'}
Role#1:{'withdraw', 'setTime'}
Time cost: 4.895964622497559
Security Policy:
Policy#0: withdraw setTime -> startTime endTime via functions setTime
total timecost: 320.3928487300873 seconds
